src/main/java/com/scylladb/jmx/metrics/RegistrationChecker.java

```diff
@@ -0,0 +1,53 @@
+package com.scylladb.jmx.metrics;
+
+import static com.scylladb.jmx.metrics.RegistrationMode.Remove;
+import static com.scylladb.jmx.metrics.RegistrationMode.Wait;
+import static java.util.EnumSet.allOf;
+import static java.util.EnumSet.of;
+
+import java.net.UnknownHostException;
+import java.util.EnumSet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.management.OperationsException;
+
+import com.scylladb.jmx.api.APIClient;
+import com.sun.jmx.mbeanserver.JmxMBeanServer;
+
+@SuppressWarnings("restriction")
+public abstract class RegistrationChecker {
+	private final Lock lock = new ReentrantLock();
+
+	public static final EnumSet<RegistrationMode> REMOVE_NO_WAIT = of(Remove);
+	public static final EnumSet<RegistrationMode> ADD_AND_REMOVE = allOf(RegistrationMode.class);
+
+	public final void reap(APIClient client, JmxMBeanServer server) throws OperationsException, UnknownHostException {
+		check(client, server, REMOVE_NO_WAIT);
+	}
+
+	public final void check(APIClient client, JmxMBeanServer server) throws OperationsException, UnknownHostException {
+		check(client, server, ADD_AND_REMOVE);
+	}
+
+	public final void check(APIClient client, JmxMBeanServer server, EnumSet<RegistrationMode> mode)
+			throws OperationsException, UnknownHostException {
+		if (!lock.tryLock() && mode.contains(Wait)) {

```

Comment:
If `tryLock` fails, but we're not allowed to wait, what is supposed to happen? Right now, we perform `doCheck` without the lock (which I assume is safe?), but we also call `unlock()`...